File: test_scripts/test_script.ts
import * as crypto from 'crypto';
import { PublicKey } from '@solana/web3.js';
import bs58 from 'bs58';

async function generateTestHash() {
    const testString = "this is a test";

    // Convert to bytes using UTF-8 encoding
    const testBytes = Buffer.from(testString, 'utf8');

    // Compute SHA256 hash
    const sha256Hash = crypto.createHash('sha256').update(testBytes).digest();

    // Convert SHA256 hash to Base58
    const base58Hash = bs58.encode(sha256Hash);

    console.log("Test String:", testString);
    console.log("Test Hash (Base58):", base58Hash);
}

async function generateSeedHash() {
    const seedPreamble = "pastel_tx_status_report";
    const txid = "9930511c526808e6849a25cb0eb6513f729c2a71ec51fbca084d7c7e4a8dea2f";
    const rewardAddress = new PublicKey("1111111QLbz7JHiBTspS962RLKV8GndWFwiEaqKM");

    // Concatenate strings
    const preimageString = seedPreamble + txid + rewardAddress.toString();

    // Convert to bytes using UTF-8 encoding
    const preimageBytes = Buffer.from(preimageString, 'utf8');

    // Compute hash
    const seedHash = crypto.createHash('sha256').update(preimageBytes).digest();

    console.log("Preimage String:", preimageString);
    console.log("Seed Hash:", seedHash.toString('hex'));

    // Generate PDA
    const programId = new PublicKey("11111111111111111111111111111111");
    const [pda, _] = await PublicKey.findProgramAddress([seedHash], programId);

    console.log("Generated PDA:", pda.toString());
}

generateTestHash();

generateSeedHash();

File: test_scripts/test_script.rs
use solana_program::{
    pubkey::Pubkey,

};

use anchor_lang::solana_program::hash::{hash, Hash};


fn main() {
    let seed_preamble = "pastel_tx_status_report";
    let txid = "9930511c526808e6849a25cb0eb6513f729c2a71ec51fbca084d7c7e4a8dea2f";
    let reward_address = Pubkey::new_unique(); // For demonstration, using a unique pubkey

    let seed_hash = create_seed(seed_preamble, txid, &reward_address);
    println!("String Seed Bytes: {:?}", seed_preamble.as_bytes());
    println!("TXID Seed Bytes: {:?}", txid.as_bytes());
    println!("Public Key Seed Bytes: {:?}", reward_address.as_ref());
    
    println!("Seed Preamble: {}", seed_preamble);
    println!("Transaction ID: {}", txid);
    println!("Reward Address: {}", reward_address);
    println!("Seed Hash: {:?}", seed_hash);
    println!("Generated PDA: {:?}", Pubkey::create_program_address(&[&seed_hash.as_ref()], &Pubkey::new_unique())); // Unique program ID for demonstration

}

fn create_seed(seed_preamble: &str, txid: &str, reward_address: &Pubkey) -> Hash {
    let mut preimage = Vec::new();
    preimage.extend_from_slice(seed_preamble.as_bytes());
    preimage.extend_from_slice(txid.as_bytes());
    preimage.extend_from_slice(reward_address.as_ref());
    hash(&preimage)
}
